bus-mapping/src/circuit_input_builder/input_state_ref.rs

```diff
@@ -824,6 +824,84 @@ impl<'a> CircuitInputStateRef<'a> {
         Ok(())
     }
 
+    /// Bus mapping for the RestoreContextGadget as used in RETURN.
+    // TODO: unify this with restore context bus mapping for STOP.
+    // TODO: unify this with the `handle return function above.`
+    pub fn handle_restore_context(
+        &mut self,
+        steps: &[GethExecStep],
+        exec_step: &mut ExecStep,
+    ) -> Result<(), Error> {
+        let call = self.call()?.clone();
+        let caller = self.caller()?.clone();
+        self.call_context_read(
+            exec_step,
+            call.call_id,
+            CallContextField::CallerId,
+            caller.call_id.into(),
+        );
+
+        let geth_step = &steps[0];
+        let geth_step_next = &steps[1];
+        let caller_gas_left = geth_step_next.gas.0 - geth_step.gas.0;
+        for (field, value) in [
+            (CallContextField::IsRoot, (caller.is_root as u64).into()),
+            (
+                CallContextField::IsCreate,
+                (caller.is_create() as u64).into(),
+            ),
+            (CallContextField::CodeHash, caller.code_hash.to_word()),
+            (CallContextField::ProgramCounter, geth_step_next.pc.0.into()),
+            (
+                CallContextField::StackPointer,
+                geth_step_next.stack.stack_pointer().0.into(),
+            ),
+            (CallContextField::GasLeft, caller_gas_left.into()),
+            (
+                CallContextField::MemorySize,
+                geth_step_next.memory.word_size().into(),
+            ),
+            (
+                CallContextField::ReversibleWriteCounter,
+                self.caller_ctx()?.reversible_write_counter.into(),
+            ),
+        ] {
+            self.call_context_read(exec_step, caller.call_id, field, value);
+        }
+
+        let [last_callee_return_data_offset, last_callee_return_data_length] = match geth_step.op {
```

Comment:
Why is the length `2` for STOP and RETURN/REVERT(offset, 0) ?
Shouldn't it be `0`?

Reply:
It is 0? `[Word::zero(); 2]` means `[Word::zero(), Word::zero()]` and not `[Word::zero(), Word::from(2)]`.

Reply:
Oh, sorry for the confusion! :flushed: